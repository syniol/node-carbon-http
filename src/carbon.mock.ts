import { HttpStatusCode, NodeRequestClient } from './http'
import { Buffer } from 'buffer'
import { ClientRequest, IncomingMessage as InSecureIncomingMessage } from 'http'


export default class CarbonHTTPMock {
  public readonly client: NodeRequestClient

  public constructor(arg: string | Error, statusCode?: HttpStatusCode) {
    this.client = (_, callback) => {
      if (typeof callback === 'function') {
        const res = (<unknown>{
          on: (e: 'data' | 'error' | 'end', listener: (data?: any) => {}) => {
            switch (e) {
              case 'data':
                if (typeof arg === 'string') {
                  listener(Buffer.from(arg, 'utf8'))
                }
                break
              case 'error':
                if (arg instanceof Error) {
                  listener(arg)
                }
                break
              default:
                listener()
            }

            return this
          },
          statusCode:
            statusCode ||
            (
              arg instanceof Error
              ? HttpStatusCode.BAD_REQUEST
              : HttpStatusCode.OK
            ),
        }) as InSecureIncomingMessage

        callback(res)
      }

      return (<unknown>{
        write: () => true,
        end: () => undefined,
      }) as ClientRequest
    }
  }
}
