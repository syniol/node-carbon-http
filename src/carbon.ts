
import { URL } from 'url'
import { TextEncoder } from 'util'
import { request as InSecureRequest } from 'http'
import { request as SecureRequest } from 'https'

import {
  NodeRequestClient,
  NodeRequestOption,
  CarbonHttpRequestOption,
  CarbonHttpResponse,
  HttpStatusCode,
  HttpMethod,
  HttpProtocol,
} from './http'


function response(
  dataBlocks: Uint8Array[],
  status: Readonly<number>
): Readonly<CarbonHttpResponse> {
  const result: Readonly<string> = Buffer
    .concat(dataBlocks)
    .toString()

  return {
    status: status,
    json(): any {
      return JSON.parse(result)
    },
    text(): string {
      return result
    },
  }
}

export function Request(
  url: Readonly<string>,
  opt?: CarbonHttpRequestOption,
  clientService?: NodeRequestClient | any,
): Promise<Readonly<CarbonHttpResponse>> {
  const urlAPI = new URL(url)
  const nodeReqOpt: NodeRequestOption = {
    method: opt?.method || HttpMethod.GET,
    path: `${urlAPI.pathname}${urlAPI.search || ''}`,
    body: new TextEncoder().encode(opt?.body),
    port: opt?.port ? opt.port : urlAPI.port,
  }

  if (opt?.headers) {
    nodeReqOpt.headers = opt.headers;
  }

  let client = InSecureRequest;
  if (clientService) {
    client = clientService;
  }

  if (urlAPI.protocol === HttpProtocol.SecureHTTP) {
    if (!clientService) {
      client = SecureRequest
    }

    nodeReqOpt.hostname = urlAPI.hostname;
  } else {
    nodeReqOpt.host = urlAPI.host;
  }

  return new Promise((resolve, reject) => {
    const req = client(nodeReqOpt, (res) => {
      const dataCollection: Uint8Array[] = []
      res.on('data', (data: Uint8Array) => {
        dataCollection.push(data)
      })

      res.on('error', (err: any) => {
        reject(err)
      })

      res.on('end', () => {
        resolve(
          response(
            dataCollection,
            res.statusCode || HttpStatusCode.OK
          ),
        )
      })
    })

    req.write(opt?.body || '')

    req.end()
  })
}
