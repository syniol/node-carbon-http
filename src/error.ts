export interface CarbonError extends Error {
  code?: number
}

export function NewCarbonError(
  msg: string,
  name?: string,
  stack?: string,
  code?: number,
): Readonly<CarbonError> {
  const err: CarbonError = {
    name: name || 'CarbonHttpException',
    message: msg,
  }

  if (stack) {
    err.stack = stack
  }

  if (code) {
    err.code = code
  }

  return err
}
