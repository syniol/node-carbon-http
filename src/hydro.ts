
import { URL } from 'url'
import { TextEncoder } from 'util'
import { request as InSecureRequest } from 'http'
import { request as SecureRequest } from 'https'

import {
  RequestType,
  HttpMethod,
  HttpRequestContext,
  HttpResponse,
  HttpStatusCode
} from './http'


export default class HydroHTTP {
  readonly #forcedClient: boolean;
  #client: RequestType

  public constructor(client?: RequestType) {
    this.#forcedClient = false;
    this.#client = InSecureRequest

    if (client) {
      this.#client = client
      this.#forcedClient = true;
    }
  }

  private response(
    dataBlocks: Uint8Array[],
    status: number
  ): Readonly<HttpResponse> {
    const result = Buffer.concat(dataBlocks).toString()

    return {
      status: status,
      json(): any {
        return JSON.parse(result)
      },
      text(): string {
        return result
      },
    }
  }

  public request(
    url: string,
    context?: HttpRequestContext,
  ): Promise<HttpResponse> {
    const urlObject = new URL(url)
    const opt = {
      method: context?.method || HttpMethod.GET,
      headers: context?.headers,
      path: `${urlObject.pathname}${urlObject.search || ''}`,
      body: new TextEncoder().encode(context?.body),
      port: context?.port ? context?.port : urlObject.port,
    }

    if (urlObject?.protocol === 'https:') {
      if (!this.#forcedClient) {
        this.#client = SecureRequest
      }

      Object.assign(opt, { hostname: urlObject.hostname })
    } else {
      Object.assign(opt, { host: urlObject.host })
    }

    return new Promise((resolve, reject) => {
      const req = this.#client(opt, (res) => {
        const dataCollection: Uint8Array[] = []
        res.on('data', (data: Uint8Array) => {
          dataCollection.push(data)
        })

        res.on('error', (err: any) => {
          reject(err)
        })

        res.on('end', () => {
          resolve(
            this.response(
              dataCollection,
              res?.statusCode || HttpStatusCode.OK
            ),
          )
        })
      })

      req.write(context?.body || '')

      req.end()
    })
  }
}
