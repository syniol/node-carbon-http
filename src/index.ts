export * from './http'
export * from './carbon'
export * from './error'
