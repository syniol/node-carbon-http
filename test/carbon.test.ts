import { Request, CarbonHttpResponse } from '../src'


describe('Request Test', () => {
  const sut = Request

  describe('given GET request is made', () => {
    let actual: CarbonHttpResponse<any>;

    beforeAll(async () => {
      actual = await sut(
        'https://api.github.com/users/syniol',
        {
          headers: {
            'User-Agent': 'PostmanRuntime/7.26.5',
            'Accept': 'application/json',
          }
        }
      )
    })

    it('should bring a response back in JSON format', () => {
      expect(actual.json()).toMatchSnapshot({
        node_id: expect.any(String),
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })

    it('should bring a response back in String format', () => {
      expect(actual.text()).toContain('"login":"syniol"')
    })

    it('should bring a response back in String format', () => {
      expect(actual.headers).toMatchObject(expect.objectContaining({
        'content-length': expect.any(String),
      }))
    })
  })
})
