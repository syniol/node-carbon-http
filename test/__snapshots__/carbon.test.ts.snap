// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Test given GET request is made should bring a response back in JSON format 1`] = `
Object {
  "avatar_url": "https://avatars.githubusercontent.com/u/68777073?v=4",
  "bio": "Software Delivery and Consultancy based in London",
  "blog": "https://syniol.com",
  "company": "Syniol Limited",
  "created_at": Any<String>,
  "email": null,
  "events_url": "https://api.github.com/users/syniol/events{/privacy}",
  "followers": 0,
  "followers_url": "https://api.github.com/users/syniol/followers",
  "following": 0,
  "following_url": "https://api.github.com/users/syniol/following{/other_user}",
  "gists_url": "https://api.github.com/users/syniol/gists{/gist_id}",
  "gravatar_id": "",
  "hireable": null,
  "html_url": "https://github.com/syniol",
  "id": 68777073,
  "location": "London",
  "login": "syniol",
  "name": "Syniol Limited",
  "node_id": Any<String>,
  "organizations_url": "https://api.github.com/users/syniol/orgs",
  "public_gists": 0,
  "public_repos": Any<Number>,
  "received_events_url": "https://api.github.com/users/syniol/received_events",
  "repos_url": "https://api.github.com/users/syniol/repos",
  "site_admin": false,
  "starred_url": "https://api.github.com/users/syniol/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/syniol/subscriptions",
  "twitter_username": "syniol",
  "type": "User",
  "updated_at": Any<String>,
  "url": "https://api.github.com/users/syniol",
  "user_view_type": "public",
}
`;
